jaxopt/_src/lbfgs.py

```diff
@@ -0,0 +1,286 @@
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Limited-memory BFGS"""
+
+from typing import Any
+from typing import Callable
+from typing import NamedTuple
+from typing import Optional
+from typing import Union
+
+from dataclasses import dataclass
+
+import jax
+import jax.numpy as jnp
+
+from jaxopt._src import base
+from jaxopt._src.line_search import BacktrackingLineSearch
+from jaxopt.tree_util import tree_vdot
+from jaxopt.tree_util import tree_scalar_mul
+from jaxopt.tree_util import tree_add_scalar_mul
+from jaxopt.tree_util import tree_sub
+from jaxopt.tree_util import tree_l2_norm
+
+
+def ihp_body_right(q, tup):
+  s, y, rho = tup
+  alpha = rho * tree_vdot(s, q)
+  q = tree_add_scalar_mul(q, -alpha, y)  # q = q - alpha * y
+  return q, alpha
+
+
+def ihp_body_left(r, tup):
+  s, y, rho, alpha = tup
+  beta = rho * tree_vdot(y, r)
+  r = tree_add_scalar_mul(r, alpha - beta, s)  # r = r + (alpha - beta) * s
+  return r, beta
+
+
+def inv_hessian_product(s_history: jnp.ndarray,
+                        y_history: jnp.ndarray,
+                        rho_history: jnp.ndarray,
+                        v: jnp.ndarray,
+                        H_init_matvec: Callable=None):
+  """Product between an approximate Hessian inverse and a vector v.
+
+  Vectors s[k] or y[k] that are all-zero are ignored.
+
+  The notation follows the reference below.
+
+  Args:
+    s_history: past iterate differences, i.e., s[k] = x[k+1] - x[k].
+    y_history: past gradient differences, i.e., y[k] = g[k+1] - g[k].
+    rho_history: rho[k] = 1. / vdot(s[k], y[k])
+    v: array to multiply with.
+    H_init_matvec: multiplication with the initial inverse Hessian
+      approximation.
+
+  Reference:
+    Jorge Nocedal and Stephen Wright.
+    Numerical Optimization, second edition.
+    Algorithm 7.4 (page 178).
+  """
+  if s_history.shape != y_history.shape:
+    raise ValueError("s_history.shape != y_history.shape")
+
+  # Compute right part.
+  q, alpha = jax.lax.scan(ihp_body_right,
+                          v,
+                          (s_history, y_history, rho_history),
+                          reverse=True)
+
+  # Compute center.
+  if H_init_matvec is not None:
+    r = H_init_matvec(q)
+  else:
+    r = q
+
+  # Compute left part.
+  r, beta = jax.lax.scan(ihp_body_left,
+                         r,
+                         (s_history, y_history, rho_history, alpha))
+
+  return r
+
+
+def update_history(history, new_value):
+  """Shift past elements to the left and add new elements at the end."""
+  return jnp.roll(history, -1, axis=0).at[-1].set(new_value)
```

Comment:
In principle, this means each update requires a full copy of the stored history vectors. My guess is that it could be significantly more efficient to treat history as a rolling buffer, and only setting a single vector. Maybe worth putting down as a "TODO".

Reply:
I left a todo for now. How would you implement this in practice?

Reply:
The main difference is that the scans inside `inv_hessian_product_leaf` would need to be over a loop index (something like `(time_index - jnp.arange(history_size)[::-1]) % history_size`) with `history` passed in via a closure.

Reply:
That's a good idea! I'm kind of sad to remove the scan-based loop as it was pretty elegant but I agree that avoiding memory copies could be very important, especially when working with huge neural networks. I'll tackle that as a separate PR.